plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/RequestMetricCollector.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.repositories.azure;
+
+/**
+ * A listener that can be hooked into any Azure Storage request allowing metric collection
+ */
+interface RequestMetricCollector {
+
+    /**
+     * Used to collect metrics after a successful request/response cycle
+     *
+     * @param requestMethod HTTP request method
+     */
+    void collectMetrics(String requestMethod);
+
+    /**
+     * Used to collect metrics after a failed request/response cycle
+     *
+     * @param requestMethod the HTTP request method
+     * @param statusCode the response status code
+     */
+    default void collectMetricsForFailedRequest(String requestMethod, int statusCode) {}
+
+    default boolean isEnabled() {
```

Comment:
If we kept this in `AzureBlobStore` as suggested above then this and `NO_OP` wouldn't be needed I think, making things much simpler.